spec/firebase_app_distribution_api_client_spec.rb

```diff
@@ -128,7 +127,99 @@
   end
 
   describe '#upload' do
-    # Empty for now
+    let(:max_polling_retries) { Fastlane::Client::FirebaseAppDistributionApiClient::MAX_POLLING_RETRIES }
+    let(:polling_interval_seconds) { Fastlane::Client::FirebaseAppDistributionApiClient::POLLING_INTERVAL_SECONDS }
+    let(:upload_status_response_success) do
+      UploadStatusResponse.new(
+        { status: "SUCCESS",
+          release: { id: "release_id" } }
+      )
+    end
+    let(:upload_status_response_in_progress) do
+      UploadStatusResponse.new(
+        { status: "IN_PROGRESS",
+          release: {} }
+      )
+    end
+    let(:upload_status_response_error) do
+      UploadStatusResponse.new(
+        { status: "ERROR",
+          release: {} }
+      )
+    end
+
+    it 'skips the upload step if the binary has already been uploaded' do
+      expect(api_client).to receive(:get_upload_token)
+        .with("app_id", fake_binary_path)
+        .and_return("upload_token")
+      expect(api_client).to receive(:get_upload_status)
+        .with("app_id", "upload_token")
+        .and_return(upload_status_response_success)
+      expect(api_client).not_to(receive(:get_upload_status))
+
+      release_id = api_client.upload("app_id", fake_binary_path)
+      expect(release_id).to eq("release_id")
+    end
+
+    it 'uploads the app binary then returns the release_id' do
+      expect(api_client).to receive(:get_upload_token)
+        .with("app_id", fake_binary_path)
+        .and_return("upload_token")
+      expect(api_client).to receive(:get_upload_status)
+        .with("app_id", "upload_token")
+        .and_return(upload_status_response_error)
+      expect(api_client).to receive(:upload_binary)
+        .with("app_id", fake_binary_path)
+      expect(api_client).to receive(:get_upload_status)
+        .with("app_id", "upload_token")
+        .and_return(upload_status_response_success)
```

Comment:
Since you're expecting :get_upload_status to get called twice, can you expect the order of the calls as well?

Also, I think this might be a thing: `expect(thing).to receive(:method).and_return(first_call_result, second_call_result, third_call_result)`

Reply:
+1 Nice find Lee!

Reply:
Nice! I was wondering if there was a way of writing them like that. I believe they are expectations are ordered by how they're written

Reply:
Nevermind, they're not but you can make them: https://til.hashrocket.com/posts/ex4n9iqwnw-message-order-constraint-in-rspec